File: Java/Ema/Src/main/java/interface/com/thomsonreuters/ema/access/SeriesEntry.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright Thomson Reuters 2015. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.thomsonreuters.ema.access;


/**
 * SeriesEntry represents an entry of Series.
 * <br>SeriesEntry associates entry's data and its data type.
 */
public interface SeriesEntry
{
	/**
	 * Returns the DataType of the entry's load. 
	 * <br>Return of {@link com.thomsonreuters.ema.access.DataType.DataTypes#ERROR}
	 * signifies error while extracting content of load.
	 * 
	 * @return data type of the contained object
	 */
	public int loadType();
	
	/**
	 * Returns a string representation of the class instance.
	 * 
	 * @return string representation of the class instance
	 */
	public String toString();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.ReqMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.ReqMsg} class reference to contained object
	 */
	public ReqMsg reqMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.RefreshMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.RefreshMsg} class reference to contained entry's load object
	 */
	public RefreshMsg refreshMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.UpdateMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.UpdateMsg} class reference to contained entry's load object
	 */
	public UpdateMsg updateMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.StatusMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.StatusMsg} class reference to contained entry's load object
	 */
	public StatusMsg statusMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.PostMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.PostMsg} class reference to contained entry's load object
	 */
	public PostMsg postMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.AckMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.AckMsg} class reference to contained entry's load object
	 */
	public AckMsg ackMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.GenericMsg}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.GenericMsg} class reference to contained entry's load object
	 */
	public GenericMsg genericMsg();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.FieldList}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.FieldList} class reference to contained entry's load object
	 */
	public FieldList fieldList();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.ElementList}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.ElementList} class reference to contained entry's load object
	 */
	public ElementList elementList();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.Map}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.Map} class reference to contained entry's load object
	 */
	public Map map();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.Vector}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.Vector} class reference to contained entry's load object
	 */
	public Vector vector();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.Series}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.Series} class reference to contained entry's load object
	 */
	public Series series();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.FilterList}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.FilterList} class reference to contained entry's load object
	 */
	public FilterList filterList();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.OmmOpaque}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.OmmOpaque} class reference to contained entry's load object
	 */
	public OmmOpaque opaque();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.OmmXml}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.OmmXml} class reference to contained entry's load object
	 */
	public OmmXml xml();
	
	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.OmmAnsiPage}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.OmmAnsiPage} class reference to contained entry's load object
	 */
	public OmmAnsiPage ansiPage();
	
	/**
	 * Returns the contained Data based on the DataType.
	 * 
	 * @return Data class reference to contained entry's load object
	 */
	public Data load();
	
	/**
	 * Returns Error.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.thomsonreuters.ema.access.OmmError}
	 * 
	 * @return {@link com.thomsonreuters.ema.access.OmmError} class reference to contained entry's load object
	 */
	public OmmError error();
	
	/** Adds a ReqMsg type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 * 
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry reqMsg(ReqMsg value);
	
	/** Adds a RefreshMsg type of OMM data to the SeriesEntry.
	  * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry refreshMsg(RefreshMsg value);

	/** Adds a StatusMsg type of OMM data to the SeriesEntry.
	   * <br>All entries must have same complex data type.
	  * 
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry statusMsg(StatusMsg value);
	
	/** Adds a UpdateMsg type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry updateMsg(UpdateMsg value);
	
	/** Adds a PostMsg type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry postMsg(PostMsg value);
	
	/** Adds a AckMsg type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry ackMsg(AckMsg value);
	
	/** Adds a GenericMsg type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry genericMsg(GenericMsg value);
	
	/** Adds a FieldList type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *   
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry fieldList(FieldList value);
	
	/** Adds a ElementList type of OMM data to the SeriesEntry.
	 * <br>All entries  must have same complex data type.
	 *   
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry elementList(ElementList value);
	
	/** Adds a Map type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *   
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry map(Map value);
	
	/** Adds a Vector type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *   
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry vector(Vector value);
	
	/** Adds a Series type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry series(Series value);
	
	/** Adds a FilterList type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry filterList(FilterList value);
	
	/** Adds a OmmOpaque type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry opaque(OmmOpaque value);
	
	/** Adds a OmmXml type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *  
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 * 
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry xml(OmmXml value);
	
	/** Adds a OmmAnsiPage type of OMM data to the SeriesEntry.
	 * <br>All entries must have same complex data type.
	 *   
	 * @throws OmmInvalidUsageException if an error is detected (exception will specify the cause of the error)
	 *
	 * @param value specifies complex type associated with this entry
	 * @return reference to this object
	 */
	public SeriesEntry ansiPage(OmmAnsiPage value);
}

File: Cpp-C/Ema/Src/Access/Impl/SeriesEntry.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2015,2019-2020,2024 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

#include "SeriesEntry.h"
#include "ExceptionTranslator.h"
#include "SeriesDecoder.h"
#include "FieldList.h"
#include "ElementList.h"
#include "Map.h"
#include "Vector.h"
#include "Series.h"
#include "FilterList.h"
#include "ReqMsg.h"
#include "RefreshMsg.h"
#include "UpdateMsg.h"
#include "StatusMsg.h"
#include "PostMsg.h"
#include "AckMsg.h"
#include "GenericMsg.h"
#include "OmmError.h"
#include "OmmAnsiPage.h"
#include "OmmXml.h"
#include "OmmJson.h"
#include "OmmOpaque.h"
#include "Utilities.h"
#include "OmmInvalidUsageException.h"

using namespace refinitiv::ema::access;

extern const EmaString& getDTypeAsString( DataType::DataTypeEnum dType );

SeriesEntry::SeriesEntry() :
 _pDecoder( 0 ),
 _pLoad( 0 ),
 _toString()
{
}

SeriesEntry::~SeriesEntry()
{
}

DataType::DataTypeEnum SeriesEntry::getLoadType() const
{
	return _pLoad->getDataType();
}

const Data& SeriesEntry::getLoad() const
{
	return *_pLoad;
}

const OmmError& SeriesEntry::getError() const
{
	if ( _pLoad->getDataType() != DataType::ErrorEnum )
	{
		EmaString temp( "Attempt to getError() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const OmmError&>( *_pLoad );
}

const ElementList& SeriesEntry::getElementList() const
{
	if ( _pLoad->getDataType() != DataType::ElementListEnum )
	{
		EmaString temp( "Attempt to getElementList() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const ElementList&>( *_pLoad );
}

const FieldList& SeriesEntry::getFieldList() const
{
	if ( _pLoad->getDataType() != DataType::FieldListEnum )
	{
		EmaString temp( "Attempt to getFieldList() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const FieldList&>( *_pLoad );
}

const FilterList& SeriesEntry::getFilterList() const
{
	if ( _pLoad->getDataType() != DataType::FilterListEnum )
	{
		EmaString temp( "Attempt to getFilterList() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const FilterList&>( *_pLoad );
}

const Map& SeriesEntry::getMap() const
{
	if ( _pLoad->getDataType() != DataType::MapEnum )
	{
		EmaString temp( "Attempt to getMap() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const Map&>( *_pLoad );
}

const Series& SeriesEntry::getSeries() const
{
	if ( _pLoad->getDataType() != DataType::SeriesEnum )
	{
		EmaString temp( "Attempt to getSeries() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const Series&>( *_pLoad );
}

const Vector& SeriesEntry::getVector() const
{
	if ( _pLoad->getDataType() != DataType::VectorEnum )
	{
		EmaString temp( "Attempt to getVector() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const Vector&>( *_pLoad );
}

const ReqMsg& SeriesEntry::getReqMsg() const
{
	if ( _pLoad->getDataType() != DataType::ReqMsgEnum )
	{
		EmaString temp( "Attempt to getReqMsg() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const ReqMsg&>( *_pLoad );
}

const RefreshMsg& SeriesEntry::getRefreshMsg() const
{
	if ( _pLoad->getDataType() != DataType::RefreshMsgEnum )
	{
		EmaString temp( "Attempt to getRefreshMsg() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const RefreshMsg&>( *_pLoad );
}

const UpdateMsg& SeriesEntry::getUpdateMsg() const
{
	if ( _pLoad->getDataType() != DataType::UpdateMsgEnum )
	{
		EmaString temp( "Attempt to getUpdateMsg() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const UpdateMsg&>( *_pLoad );
}

const StatusMsg& SeriesEntry::getStatusMsg() const
{
	if ( _pLoad->getDataType() != DataType::StatusMsgEnum )
	{
		EmaString temp( "Attempt to getStatusMsg() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const StatusMsg&>( *_pLoad );
}

const PostMsg& SeriesEntry::getPostMsg() const
{
	if ( _pLoad->getDataType() != DataType::PostMsgEnum )
	{
		EmaString temp( "Attempt to getRespMsg() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const PostMsg&>( *_pLoad );
}

const AckMsg& SeriesEntry::getAckMsg() const
{
	if ( _pLoad->getDataType() != DataType::AckMsgEnum )
	{
		EmaString temp( "Attempt to getAckMsg() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const AckMsg&>( *_pLoad );
}

const GenericMsg& SeriesEntry::getGenericMsg() const
{
	if ( _pLoad->getDataType() != DataType::GenericMsgEnum )
	{
		EmaString temp( "Attempt to getGenericMsg() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const GenericMsg&>( *_pLoad );
}

const OmmOpaque& SeriesEntry::getOpaque() const
{
	if ( _pLoad->getDataType() != DataType::OpaqueEnum )
	{
		EmaString temp( "Attempt to getOpaque() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const OmmOpaque&>( *_pLoad );
}

const OmmXml& SeriesEntry::getXml() const
{
	if ( _pLoad->getDataType() != DataType::XmlEnum )
	{
		EmaString temp( "Attempt to getXml() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const OmmXml&>( *_pLoad );
}

const OmmJson& SeriesEntry::getJson() const
{
	if ( _pLoad->getDataType() != DataType::JsonEnum )
	{
		EmaString temp( "Attempt to getJson() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const OmmJson&>( *_pLoad );
}

const OmmAnsiPage& SeriesEntry::getAnsiPage() const
{
	if ( _pLoad->getDataType() != DataType::AnsiPageEnum )
	{
		EmaString temp( "Attempt to getAnsiPage() while actual data type is " );
		temp += getDTypeAsString( _pLoad->getDataType() );
		throwIueException( temp, OmmInvalidUsageException::InvalidOperationEnum );
	}

	return static_cast<const OmmAnsiPage&>( *_pLoad );
}

const EmaString& SeriesEntry::toString() const
{
	_toString.clear().append( "SeriesEntry dataType=\"" ).append( getDTypeAsString( _pDecoder->getLoad().getDataType() ) ).append( "\"\n" );
	_toString += _pDecoder->getLoad().toString( 1 );
	addIndent( _toString, 0 ).append( "SeriesEntryEnd\n" );

	return _toString;
}
